pkg/nginxprocess/process.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+// Package nginxprocess contains utilities for working with OS-level NGINX processes.
+package nginxprocess
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/shirou/gopsutil/v4/process"
+)
+
+// Process contains a snapshot of read-only data about an OS-level NGINX process. Create using [List] or [Find].
+type Process struct {
+	Created time.Time
+	Name    string
+	Cmd     string
+	Exe     string // path to the executable
+	Status  string // process status, only present if this process was created using [WithStatus]
+	PID     int32
+	PPID    int32 // parent PID
+}
+
+// IsWorker returns true if the process is a NGINX worker process.
+func (p *Process) IsWorker() bool { return strings.Contains(p.Cmd, "worker") }
+
+// IsShuttingDown returns true if the process is shutting down. This can identify workers that are in the process of a
+// graceful shutdown. See [changing NGINX configuration] for more details.
+//
+// [changing NGINX configuration]: https://nginx.org/en/docs/control.html#reconfiguration
+func (p *Process) IsShuttingDown() bool { return strings.Contains(p.Cmd, "is shutting down") }
+
+// IsHealthy uses Status flags to judge process health. Only works on processes created using [WithStatus].
+func (p *Process) IsHealthy() bool {
+	return p.Status != "" && !strings.Contains(p.Status, process.Zombie)
+}
+
+type options struct {
+	loadStatus bool
+}
+
+// Option customizes how processes are gathered from the OS.
```

Comment:
Maybe learning for me, the `interface` for `apply` seems extraneous to me. Why not just 

```
type Option func(opts *options)

func WithStatus(v bool) Option {
	return func(opts *options) {
		opts.loadStatus = v
	}
}
```

Reply:
this is the "functional options" implementation we settled on NGINXaaS. It does take more code, but sets a stronger boundary between exported and unexported symbols. This is a similar approach used by [opentelemetry](https://github.com/open-telemetry/opentelemetry-go/blob/c862d2d69cae9f4af2fe8ab244a3b5054d223b55/trace/config.go#L178).

The downside of `type Option func(opts *options)` is that it exposes the unexported `options` type as part of the exported `Option` type. There's not much a caller can do with that information, but it's exposing details we might want to change in the future without having to think about compatibility.

The other (more hypothetical) benefit is flexibility. In the unlikely event this package grows a lot of more features, we could add another option interface (e.g maybe we split into `FindOption` and `ListOption`). `WithStatus` could return a type that implements two interfaces, but that's not possible with `func` types.

If this code were in an `internal` package, I'd probably just use `type Option func(opts *options)`, as all the callers are known. Since this is a public codebase in a package meant to be consumed by other software, I figured I'd err on the side of paranoia.